source/Engine.cpp

```diff
@@ -870,9 +866,29 @@ void Engine::Step(bool isActive)
 	// Draw crosshairs on any minables in range of the flagship's scanners.
 	double scanRange = flagship ? 100. * sqrt(flagship->Attributes().Get("asteroid scan power")) : 0.;
 	if(flagship && scanRange && !flagship->IsHyperspacing())
+	{
+		// Decide before looping whether or not to catalog asteroids.  This
+		// results in cataloging in-range asteroids roughly 3 times a second.
+		bool shouldCatalogAsteroids = !Random::Int(20);
+		shouldCatalogAsteroids &= !isAsteroidCatalogComplete;
+		bool scanComplete = true;
 		for(const shared_ptr<Minable> &minable : asteroids.Minables())
 		{
 			Point offset = minable->Position() - center;
+
+			// Autocatalog asteroid: Record that the player knows this type of asteroid is available here.
+			if(shouldCatalogAsteroids)
+			{
+				if(!Random::Int(10) && (minable->Position() - flagship->Position()).Length() <= scanRange)
+				{
+					asteroidsScanned.insert(minable->Name());
+					for(const auto &it : minable->Payload())
+						player.Harvest(it.first);
+				}
+				if(!asteroidsScanned.count(minable->Name()))
+					scanComplete = false;
+			}
+
 			if(offset.Length() > scanRange && flagship->GetTargetAsteroid() != minable)
 				continue;
```

Comment:
This already checks whether the minable is in range, so there's no need to duplicate the check if you move the logic below this if.